### DISCUSSION TITLE
Some questions about the future

### DISCUSSION THREAD

**Discussion by OldSnake22** [jrouwe/JoltPhysics#1263] [General]
Hello, 

First, thank you for this awesome physics engine @jrouwe, really appreciate the active development for it ! I will begin to use it very soon in my engine and the current features are looking great. 

In the past I have been using Nvidia physics which is also great but the fact that some features (like cloth or soft bodies) require an NVidia GPU because it is using GPU acceleration with CUDA cores is problematic as I would like to add those feature for everyone and not just NVidia users. Jolt is very good regarding this subject and that's why i'm thinking to use it instead of PhysX.

Could you tell us more about the future of Jolt ? What are the priorities right now ? Can we expect some new major features in the near future ? Are you considering fluids/particles simulation for example ? 

Thank you again and keep up your great work !

**Comment by jrouwe** [comment]
Hello,

My current focus is getting [godot-jolt](https://github.com/godot-jolt/godot-jolt) to feature parity with godot physics. This means I'm developing some new stuff for soft body physics at the moment. Besides that, many of the things I work on are triggered by questions on github.

W.r.t. fluid sims: I think the only way to do them at scale is by having GPU acceleration. GPU acceleration is completely missing from Jolt at the moment. It is something I'm interested in, but I'd probably start by making the rigid/soft body sim run on GPU. Fluid sim is pretty low on the list.

**Reply by rgbforge** [reply]
I don't know how much interest you have in this given the amount of directional shift involved, but I'm going to be working on my master's thesis, and have chosen to develop an experimental version of Jolt that uses [Kokkos](https://github.com/kokkos/kokkos) to support multiple compute backends including all mainstream GPUs + OpenMP. There's a productivity wrapper called [Matar](https://github.com/lanl/MATAR) may or may not end up being the source of the types, we'll see.

It may involve some serious rethinking of the  threading/parallelization within Jolt, so it may not be something you'd consider making native to your repo given it sounds like you have your own plans.

Assuming the conversion to supporting Kokkos goes well, I will then be developing a FLIP method using as much of the Jolt "style" as possible to reduce any overhead when integrating to Godot. The FLIP method will be the core component of the thesis, the Kokkos support is just to test and study the conceptual feasibility.

Kinda up to you whether I contribute it or turn it into a hard fork or keep the conversation open to see if you'd be interested in considering a merge later on.

Cheers!

**Reply by jrouwe** [reply]
Since my last post, I have done some investigation into fluid sims. I have built a Metal and Vulkan based rendering back end. The intention was to use it to run a fluid sim on GPU, but priorities have shifted a bit and now I'm looking into something else that also requires GPU sim. Chances are pretty high that an API to run work on the GPU will land in the coming months. My goal for Jolt is to not have any dependencies on 3rd party libraries (other than those that are available in the OS), so I'll be building an interface with implementations for DirectX/Vulkan/Metal. A dependency on Kokkos is not something that I want to accept, but if you build a good fluid sim then I'm definitively interested in integrating that in some way.

**Reply by rgbforge** [reply]
That's awesome news!

Adding a section that would compare a DX/vulkan backend would be icing on the cake, so I'll keep that in mind. The data structures and API investigation will dominate the next two semesters, so true fluid sim work won't start until Summer 2026. Given your velocity, I'll drop back by next summer for updates on that part if you haven't beaten me to it :)

In the meantime, I'll still be neck deep in Jolt, so hopefully I can contribute some benchmarking, testing, and other work along the way.

**Reply by AnuraDev** [reply]
btw ETA on any GPU based simulation demos? how is the progress in general?

**Reply by jrouwe** [reply]
I'm working in a private fork on a GPU interface. I've got one working for DX12 but still need to implement it for the other platforms. I'm also working on something that uses it, but it is a lot of work and will likely still take several months before I will push anything to the main repo. At that point there will be the possibility to run things on GPU, the thing that uses it is likely to come even later.

**Reply by AnuraDev** [reply]
awesome!

**Reply by luigicaval** [reply]
Great, wonderful news! Congratulations @jrouwe.

Just a question : on the doc of Godot 4.5 we can read
_Joint properties(https://docs.godotengine.org/en/stable/tutorials/physics/using_jolt_physics.html#joint-properties)
The current interfaces for the 3D joint nodes don't quite line up with the interface of Jolt's own joints. As such, there are a number of joint properties that are not supported, mainly ones related to configuring the joint's soft limits.
The unsupported properties are:
PinJoint3D: bias, damping, impulse_clamp
HingeJoint3D: bias, softness, relaxation
SliderJoint3D: angular_\*, \*_limit/softness, \*_limit/restitution, \*_limit/damping
ConeTwistJoint3D: bias, relaxation, softness
Generic6DOFJoint3D: *_limit_*/softness, *_limit_*/restitution, *_limit_*/damping, *_limit_*/erp
Currently a warning is emitted if you set these properties to anything but their default values._

So, these properties are significant ? I just learn the joints now, I love it! Do you complete these properties or it isn't worth ?

**Reply by jrouwe** [reply]
This is very off topic, please start a new thread.

**Reply by luigicaval** [reply]
Off topic, sorry I believed that it was included in "Some questions about the future" as I don't know if the properties cited before will be implemented or not... in the future :) Sorry, I didn't want to create with a new thread with a trivial question without much importance. But I will think  of a new question in a new thread  ;)

**Reply by jrouwe** [reply]
I've opened up a PR that exposes the compute interface: https://github.com/jrouwe/JoltPhysics/pull/1847. CI is failing at the moment, so it needs a bit more work.
